main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -44,255 +45,267 @@ object Safety {
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def): List[CompilationMessage] = visitExp(def0.impl.exp)
+  private def visitDef(def0: Def)(implicit flix: Flix): List[CompilationMessage] = {
+    val rigidityEnv = def0.spec.tparams.map(_.sym).foldLeft(RigidityEnv.empty) {
+      case (acc, e) => acc.markRigid(e)
+    }
+    visitExp(def0.impl.exp, rigidityEnv)
+  }
+
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(exp0: Expression): List[CompilationMessage] = exp0 match {
-    case Expression.Unit(_) => Nil
+  private def visitExp(exp0: Expression, rigidityEnv: RigidityEnv)(implicit flix: Flix): List[CompilationMessage] = {
 
-    case Expression.Null(_, _) => Nil
+    def vstExp(exp1: Expression): List[CompilationMessage] = exp1 match {
```

Comment:
`exp1` has shadows below. Can we do `e0` for the top expression, then `exp0` for the inner one?